File: apps/api/src/application/commands/events/create-event.command.ts
import { ICommand, ICommandHandler, CommandHandler } from '@nestjs/cqrs';
import { Inject, ForbiddenException, BadRequestException } from '@nestjs/common';
import {
  IEventRepository,
  EVENT_REPOSITORY,
} from '../../../domain/repositories/event.repository.interface';
import {
  ICoupleRepository,
  COUPLE_REPOSITORY,
} from '../../../domain/repositories/couple.repository.interface';
import { Event, EventCategory } from '../../../domain/aggregates/event.aggregate';

export class CreateEventCommand implements ICommand {
  constructor(
    public readonly userId: string,
    public readonly coupleId: string,
    public readonly title: string,
    public readonly startTime: Date,
    public readonly endTime: Date,
    public readonly category: EventCategory,
    public readonly description?: string,
  ) {}
}

@CommandHandler(CreateEventCommand)
export class CreateEventCommandHandler
  implements ICommandHandler<CreateEventCommand>
{
  constructor(
    @Inject(EVENT_REPOSITORY)
    private readonly eventRepository: IEventRepository,
    @Inject(COUPLE_REPOSITORY)
    private readonly coupleRepository: ICoupleRepository,
  ) {}

  async execute(command: CreateEventCommand): Promise<Event> {
    const { userId, coupleId, title, startTime, endTime, category, description } =
      command;

    const couple = await this.coupleRepository.findById(coupleId);
    if (!couple) {
      throw new BadRequestException('Couple not found');
    }

    if (!couple.hasUser(userId)) {
      throw new ForbiddenException('User does not belong to this couple');
    }

    const event = Event.create({
      coupleId,
      title,
      startTime,
      endTime,
      category,
      authorId: userId,
      description,
    });

    await this.eventRepository.save(event);

    return event;
  }
}

File: apps/api-kotlin/src/main/kotlin/com/couplecalendar/application/command/event/CreateEventCommand.kt
package com.couplecalendar.application.command.event

import com.couplecalendar.application.command.Command
import com.couplecalendar.application.command.CommandHandler
import com.couplecalendar.common.exception.BadRequestException
import com.couplecalendar.common.exception.ForbiddenException
import com.couplecalendar.domain.aggregate.Event
import com.couplecalendar.domain.aggregate.EventCategory
import com.couplecalendar.domain.repository.CoupleRepository
import com.couplecalendar.domain.repository.EventRepository
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.time.Instant
import java.util.UUID

data class CreateEventCommand(
    val userId: UUID,
    val coupleId: UUID,
    val title: String,
    val startTime: Instant,
    val endTime: Instant,
    val category: EventCategory,
    val description: String? = null,
    val location: String? = null
) : Command<Event>

@Component
class CreateEventCommandHandler(
    private val eventRepository: EventRepository,
    private val coupleRepository: CoupleRepository
) : CommandHandler<CreateEventCommand, Event> {

    @Transactional
    override fun handle(command: CreateEventCommand): Event {
        val couple = coupleRepository.findById(command.coupleId)
            ?: throw BadRequestException("Couple not found")

        if (!couple.hasUser(command.userId)) {
            throw ForbiddenException("User is not a member of this couple")
        }

        val event = Event.create(
            coupleId = command.coupleId,
            title = command.title,
            startTime = command.startTime,
            endTime = command.endTime,
            category = command.category,
            authorId = command.userId,
            description = command.description,
            location = command.location
        )

        eventRepository.save(event)
        return event
    }
}
